### PR TITLE
Implement backoff

### PR DESCRIPTION
**Pull request by fbielejec**

### CODE REVIEW

File: relayer/relayer/src/main.rs [hunk 1]
```diff
@@ -183,17 +190,29 @@ async fn main() -> Result<(), RelayerError> {
...
                 if tasks.is_empty() {
-                    info!("Relayer exited. Waiting {delay:?} before rebooting.");
+                    let tock = tick.elapsed();
+                    info!("Relayer exited after {tock:?}. ");
+
+                    // if relayed worked for over 60 minutes reset the delay
+                    if tock >= Duration::from_millis(600000) {
+                        delay = DEFAULT_DURATION;
+                    } else {
+                        delay = min(MAX_DURATION, delay + delay / 10);
```
[the remark this answers is not in the document]
Comment (fbielejec) [reply relayer/relayer/src/main.rs:209]: correct, numbers itself up for discussion

[the remark this answers is not in the document]
Comment (krzysztofziobro) [reply relayer/relayer/src/main.rs:209]: Would extract the minimum correct task length value to a constant (or maybe to a config parameter, but the config is getting quite big already).

Also for the values: would set max duration to 1-2 minutes.
and the task length to 10-20 minutes

[the remark this answers is not in the document]
Comment (fbielejec) [reply relayer/relayer/src/main.rs:209]: 1-2 minutes is hardly a backoff. I propose:

```
/// minimum amount of time the relayer should run healthy to reset the backoff duration to the default value
const MINIMUM_TASK_LENGHT: Duration = Duration::from_millis(600000); // 10 minutes
/// starting backoff value
const DEFAULT_BACKOFF_DURATION: Duration = Duration::from_millis(2000); // 2 seconds
/// maximal backoff value
const MAX_BACKOFF_DURATION: Duration = Duration::from_millis(600000); // 10 minutes
```

[the remark this answers is not in the document]
Comment (krzysztofziobro) [reply relayer/relayer/src/main.rs]: OK, that sounds reasonable

File: relayer/relayer/src/main.rs [hunk 2]
```diff
@@ -178,17 +189,29 @@ async fn main() -> Result<(), RelayerError> {
...
             Ok(result) => {
                 debug!("One of the core components exited gracefully due to : {result:?}, remaining: {}", &tasks.len());
 
-                // TODO: restart with backoff
                 if tasks.is_empty() {
-                    info!("Relayer exited. Waiting {delay:?} before rebooting.");
+                    let tock = tick.elapsed();
+                    info!("Relayer exited after {tock:?}. ");
+
+                    // if relayed worked for over 60 minutes reset the delay
```
Comment (JanKuczma) [inline relayer/relayer/src/main.rs:204]:
```suggestion
```
nit: maybe remove this comment (I think it's inaccurate). otherwise lgtm

**Review by JanKuczma — APPROVED** [review]

**Pull request merged**